solver/src/main.rs

```diff
@@ -192,3 +197,59 @@ async fn main() {
     serve_metrics(registry, ([0, 0, 0, 0], args.metrics_port).into());
     driver.run_forever().await;
 }
+
+async fn build_amm_artifacts(
+    sources: Vec<AmmSources>,
+    chain_id: u64,
+    settlement_contract: contracts::GPv2Settlement,
+    base_tokens: HashSet<H160>,
+    web3: web3::Web3<LoggingTransport<Http>>,
+) -> (PoolFetcher, Vec<UniswapLikeLiquidity>) {
+    let mut res = vec![];
+    // At the moment, we build uniswap artifacts for price finding.
+    // TODO - Eventually this code will move into the match statement.
+    let uniswap_router = contracts::UniswapV2Router02::deployed(&web3)
+        .await
+        .expect("couldn't load deployed uniswap router");
+    let uniswap_factory = contracts::UniswapV2Factory::deployed(&web3)
+        .await
+        .expect("couldn't load deployed uniswap router");
+    let uniswap_pair_provider = Arc::new(UniswapPairProvider {
+        factory: uniswap_factory,
+        chain_id,
+    });
+    let pool_fetcher = PoolFetcher {
```

Comment:
Maybe you can create the pool fetcher outside of this method. I don't think it's a problem to instantiate two pair providers in separate places.